arch/i386/mcount-support.c

```diff
@@ -0,0 +1,236 @@
+	asm volatile ("movsd %%xmm7, %0\n" : "=m" (ctx->xmm[7]));
+}
+
+void mcount_restore_arch_context(struct mcount_arch_context *ctx)
+{
+	asm volatile ("movsd %0, %%xmm0\n" :: "m" (ctx->xmm[0]));
+	asm volatile ("movsd %0, %%xmm1\n" :: "m" (ctx->xmm[1]));
+	asm volatile ("movsd %0, %%xmm2\n" :: "m" (ctx->xmm[2]));
+	asm volatile ("movsd %0, %%xmm3\n" :: "m" (ctx->xmm[3]));
+	asm volatile ("movsd %0, %%xmm4\n" :: "m" (ctx->xmm[4]));
+	asm volatile ("movsd %0, %%xmm5\n" :: "m" (ctx->xmm[5]));
+	asm volatile ("movsd %0, %%xmm6\n" :: "m" (ctx->xmm[6]));
+	asm volatile ("movsd %0, %%xmm7\n" :: "m" (ctx->xmm[7]));
+}
+
+unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
+{
+	struct sym *sym;
+
+	sym = &pd->dsymtab.sym[idx];
+	return sym->addr + ARCH_PLTHOOK_ADDR_OFFSET;
+}
+
+
+/*
+	For 16-byte stack-alignment, 
+	the main function stores the return address in its stack scope at prologue.
+	When the time comes for the main function to return,
+	1. restore the saved return address from stack.
+	2. After cleaning up the stack.
+	3. Put the return address at the top of the stack and return.
+	4. will be returned.
+
+	080485f8 <main>:
+	80485f8: 8d 4c 24 04           lea    0x4(%esp),%ecx
+	80485fc: 83 e4 f0              and    $0xfffffff0,%esp
+	80485ff: ff 71 fc              pushl  -0x4(%ecx)
+	8048602: 55                    push   %ebp
+	8048603: 89 e5                 mov    %esp,%ebp
+	8048605: 51                    push   %ecx
+	8048606: 83 ec 14              sub    $0x14,%esp
+	8048609: e8 02 fe ff ff        call   8048410 <mcount@plt>
+
+	... ... 
+
+	8048645: 8b 4d fc              mov    -0x4(%ebp),%ecx
+	8048648: c9                    leave
+	8048649: 8d 61 fc              lea    -0x4(%ecx),%esp
+	804864c: c3                    ret
+
+	So, in this case. The return address we want to replace with 
+	mcount_exit is in the stack scope of the main function. 
+	Non a parent located. 
+
+	we search stack for that address. 
+	we will look for it.
+	we will find it, and we will replace it. 
+	GOOD LUCK!
+*/
+unsigned long *mcount_arch_parent_location(struct symtabs *symtabs,
+                                            unsigned long *parent_loc, 
+                                            unsigned long child_ip)
+{
+	if (!search_main_ret) {
+		struct sym *parent_sym, *child_sym;
+		char *parent_name, *child_name;
+
+		const char *find_main[] = {
+			"__libc_start_main",
+			"main"
+		};
+		unsigned long ret_addr;
+		unsigned long search_ret_addr;
+
+		ret_addr = *parent_loc;
+		parent_sym = find_symtabs(symtabs, ret_addr);
+		parent_name = symbol_getname(parent_sym, ret_addr);
+		child_sym = find_symtabs(symtabs, child_ip);
+		child_name = symbol_getname(child_sym, child_ip);
+
+		// Assuming that this happens only in main..			
+		bool found_main_ret = false;
+		if (!(strcmp(find_main[0], parent_name) || 
+		      strcmp(find_main[1], child_name))) {
+			ret_addr = *parent_loc;
+			for (int i = 1; i < 5; i++ ) {
```

Comment:
You'd be better to use a symbol name instead of the magic number of "5" (isn't "4" enough?).  What about `MAX_STACK_ADJ` or something?

Reply:
OK!